ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java

```diff
@@ -65,6 +74,30 @@ public long getSize() {
   }
 
   @Override
+  public void setDigestAlgorithm(String digestAlgorithm)
+      throws NoSuchAlgorithmException {
+    if (digest != null && digest.getAlgorithm().equals(digestAlgorithm)) {
+      return;
+    }
+    digest = MessageDigest.getInstance(digestAlgorithm);
+    digestBytes = null;
+  }
+
+  @Override
+  public byte[] getDigest() {
```

Comment:
I think you could add some docs here. The interface's docs say that certain things are left to the implementation, and since this is the implementation you could add some concrete information. Like this class seems to be okay with a `getDigest()` even if the buffer is only partially consumed.